axelor-budget/src/test/java/com/axelor/apps/budget/service/TestBudgetToolsService.java

```diff
@@ -0,0 +1,299 @@
+import com.axelor.apps.base.db.Company;
+import com.axelor.apps.budget.db.Budget;
+import com.axelor.apps.budget.db.BudgetDistribution;
+import com.axelor.apps.budget.db.BudgetLevel;
+import com.axelor.apps.budget.db.BudgetStructure;
+import com.axelor.apps.budget.db.GlobalBudget;
+import com.axelor.apps.budget.db.repo.BudgetLevelRepository;
+import com.axelor.apps.budget.db.repo.BudgetRepository;
+import com.axelor.apps.budget.db.repo.GlobalBudgetRepository;
+import com.axelor.apps.budget.module.BudgetTest;
+import com.axelor.auth.db.Role;
+import com.axelor.auth.db.User;
+import com.axelor.auth.db.repo.RoleRepository;
+import com.axelor.auth.db.repo.UserRepository;
+import com.axelor.inject.Beans;
+import com.axelor.meta.loader.LoaderHelper;
+import com.axelor.studio.db.AppBudget;
+import com.axelor.studio.db.repo.AppBudgetRepository;
+import com.google.inject.Inject;
+import com.google.inject.persist.Transactional;
+import com.google.inject.servlet.RequestScoper;
+import com.google.inject.servlet.ServletScopes;
+import java.math.BigDecimal;
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+public class TestBudgetToolsService extends BudgetTest {
+
+  protected final GlobalBudgetRepository globalBudgetRepository;
+  protected final BudgetRepository budgetRepository;
+  protected final BudgetLevelRepository budgetLevelRepository;
+  protected final UserRepository userRepository;
+  protected final RoleRepository roleRepository;
+  protected final AppBudgetRepository appBudgetRepository;
+  protected final BudgetToolsService budgetToolsService;
+  protected static final LoaderHelper loaderHelper = Beans.get(LoaderHelper.class);
+
+  @Inject
+  public TestBudgetToolsService(
+      GlobalBudgetRepository globalBudgetRepository,
+      BudgetRepository budgetRepository,
+      BudgetLevelRepository budgetLevelRepository,
+      UserRepository userRepository,
+      RoleRepository roleRepository,
+      AppBudgetRepository appBudgetRepository) {
+    this.globalBudgetRepository = globalBudgetRepository;
+    this.budgetRepository = budgetRepository;
+    this.budgetLevelRepository = budgetLevelRepository;
+    this.userRepository = userRepository;
+    this.roleRepository = roleRepository;
+    this.appBudgetRepository = appBudgetRepository;
+
+    RequestScoper scope = ServletScopes.scopeRequest(Collections.emptyMap());
+    try (RequestScoper.CloseableScope ignored = scope.open()) {
+      this.budgetToolsService = Beans.get(BudgetToolsService.class);
+    }
+  }
+
+  @BeforeAll
+  static void setUp() {
+    loaderHelper.importCsv("data/base-config.xml");
+    loaderHelper.importCsv("data/budget-input.xml");
+    loaderHelper.importCsv("data/budget-template-input.xml");
+  }
+
+  @Test
+  void testCheckBudgetKeyAndRole() throws AxelorException {
```

Comment:
Please refactor this method, it seems like you should have 4 methods:

testCheckBudgetKeyAndRole_when_user_sets_are_empty()
testCheckBudgetKeyAndRole_when_user_sets_have_no_roles()
...